### PR TITLE
Soroban upgrades for simulation tests

### PR DESCRIPTION
**Pull request by ThomasBrady**

### CODE REVIEW

File: src/test/TestUtils.cpp [hunk 1]
```diff
@@ -194,6 +197,78 @@ genesis(int minute, int second)
...
+{
+    auto nodes = simulation->getNodes();
+    auto& lg = nodes[0]->getLoadGenerator();
+    auto& app = *nodes[0];
+
+    auto& complete =
+        app.getMetrics().NewMeter({"loadgen", "run", "complete"}, "run");
+    auto completeCount = complete.count();
+    // Only create an account if there are none aleady created.
+    if (app.getMetrics()
```
Comment (marta-lokhova) [inline src/test/TestUtils.cpp:212]: Why do we need to create accounts here? We could just use the root account right?

File: src/test/TestUtils.cpp [hunk 2]
```diff
@@ -194,6 +197,78 @@ genesis(int minute, int second)
...
+            .NewMeter({"loadgen", "account", "created"}, "account")
+            .count() == 0)
+    {
+        auto createAccountsLoadConfig =
+            GeneratedLoadConfig::createAccountsLoad(1, 1);
+        createAccountsLoadConfig.offset =
```
Comment (marta-lokhova) [inline src/test/TestUtils.cpp:218]: this doesn't look right: offsets are used if multiple nodes are creating accounts at the same time, so we use offset as to not overlap created accounts. Here we're only using one node, so we shouldn't need to configure offsets (same applies to other loadgen modes below)

File: src/test/TestUtils.cpp [hunk 3]
```diff
@@ -194,6 +197,78 @@ genesis(int minute, int second)
...
+            app->getLedgerManager().getLastClosedLedgerHeader().header;
+        scheduledUpgrades.mUpgradeTime =
+            VirtualClock::from_time_t(lclHeader.scpValue.closeTime);
+        scheduledUpgrades.mConfigUpgradeSetKey = upgradeSetKey;
+        app->getHerder().setUpgrades(scheduledUpgrades);
+    }
+    // Wait for upgrade to be applied
+    simulation->crankUntil(
+        [&]() {
+            auto status = app.getStatusManager().getStatusMessage(
```
Comment (marta-lokhova) [inline src/test/TestUtils.cpp:264]: This doesn't properly wait for the upgrade to go through: setUpgrades sets status message REQUIRES_UPGRADES immediately, so this call to crankUntil is essentially a no-op. This is probably why some of the tests are failing. 

I'd suggest cranking until `cfg` equals config you get from `LedgerManager.getSorobanNetworkConfig`. Basically, you need to ensure that upgrades were applied to the ledger.

**Review by marta-lokhova — CHANGES_REQUESTED** [review]

File: src/test/TestUtils.cpp (hunk 2 above)
Comment (ThomasBrady) [reply src/test/TestUtils.cpp:218]: The issue is that this function is used by tests that create accounts after calling `upgradeSorobanNetworkConfig`. If an offset is not used, the test case's subsequent account creation load generation will fail as the first `CREATE_ACCOUNT` operation will return `CREATE_ACCOUNT_ALREADY_EXIST`. The the load generation will not complete and the simulation will timeout.

File: src/test/TestUtils.cpp (hunk 1 above)
Comment (ThomasBrady) [reply src/test/TestUtils.cpp:212]: How can I use the root account? If I remove the account creation step I get: ` Account 0 must exist in the DB.`

Comment (marta-lokhova) [reply src/test/TestUtils.cpp:212]: ah I see, loadgen doesn't mark root account as "available to use" for some reason. if it did, it would simplify things since you wouldn't need account creation + offset. You could modify this path https://github.com/stellar/stellar-core/blob/ad7b1646209fcd069c067a9a7730c9b6305be667/src/simulation/LoadGenerator.cpp#L646 

Instead, it would also return root account (mRoot) if it's available. That's lesser priority though, and can be done as a follow-up.

File: src/test/TestUtils.cpp (hunk 3 above)
Comment (ThomasBrady) [reply src/test/TestUtils.cpp:264]: As I understand it, `REQUIRES_UPGRADES` is a category of status message, `setUpgrades` [ sets the contents ](https://github.com/stellar/stellar-core/blob/master/src/herder/HerderImpl.cpp#L1513) of the `REQUIRES_UPGRADES` status message to "Armed with network upgrades...". When the upgrades are applied, the `REQUIRES_UPGRADES` status message is removed. This cranks until that status message is empty. I'm fairly certain this works as I have [assertions on the upgraded values after this completes](https://github.com/stellar/stellar-core/pull/4413/files/7275997f7876adcbc806bc99772347f0f30f8b71#diff-519110afbb88598be9bd6ca994ca41c4a80840b66febd9d58765ca96955c3389R122), but I can change it to compare with the config from `LedgerManager.getSorobanNetworkConfig`.

Comment (marta-lokhova) [reply src/test/TestUtils.cpp:264]: I see. I believe relying on status message still isn't enough to tell that upgrades went through, for a few reasons:
* Core clears upgrades when it reaches consensus on a block containing upgrades, but _before_ the block is applied. This means that your crank function is potentially racing with the actual upgrade application, so we don't know if upgrades were applied when this function exits. 
* It seems a bit error-prone to rely on an empty status message: core is a state machine so it might enter a different state, and set status to something else, which would cause the simulation to time out.

I think the most reliable approach would be to rely directly on what's in the database, e.g. use `getSorobanNetworkConfig`

File: src/herder/test/HerderTests.cpp [hunk 4]
```diff
@@ -4654,31 +4656,29 @@ TEST_CASE("do not flood too many soroban transactions",
...
+
+    auto cfgUpgradeSetKey = upgradeSorobanNetworkConfig(
+        [&](SorobanNetworkConfig& cfg) {
+            setSorobanNetworkConfigForTest(cfg);
+            // Update read entries to allow flooding at most 1 tx per broadcast
+            // interval.
+            cfg.mLedgerMaxReadLedgerEntries = 40;
+            cfg.mLedgerMaxReadBytes = cfg.mTxMaxReadBytes;
+        },
+        simulation);
```
Comment (marta-lokhova) [inline src/herder/test/HerderTests.cpp:4681]: I think this test is now failing because we start flooding a little bit later (because of the upgrade), so transaction queue counts that time towards total allowed quota it can flood. You can fix this by clearing the carry-over in the transaction queue, here's the diff:
```
diff --git a/src/herder/TransactionQueue.h b/src/herder/TransactionQueue.h
index a58573a1a..4d531a68f 100644
--- a/src/herder/TransactionQueue.h
+++ b/src/herder/TransactionQueue.h
@@ -259,6 +259,14 @@ class SorobanTransactionQueue : public TransactionQueue
     }

     size_t getMaxQueueSizeOps() const override;
+#ifdef BUILD_TESTS
+    void
+    clearBroadcastCarryover()
+    {
+        mBroadcastOpCarryover.clear();
+        mBroadcastOpCarryover.resize(1, Resource::makeEmptySoroban());
+    }
+#endif

   private:
     virtual std::pair<Resource, std::optional<Resource>>
diff --git a/src/herder/test/HerderTests.cpp b/src/herder/test/HerderTests.cpp
index 03eee093e..603e57b22 100644
--- a/src/herder/test/HerderTests.cpp
+++ b/src/herder/test/HerderTests.cpp
@@ -4769,6 +4769,8 @@ TEST_CASE("do not flood too many soroban transactions",
     {
         // no broadcast right away
         REQUIRE(numBroadcast == 0);
+        tq.clearBroadcastCarryover();
+
         // wait for a bit more than a broadcast period
         // rate per period is 100 ms
         auto broadcastPeriod =
```

**Comment by marta-lokhova** [comment]
@ThomasBrady what other tests are failing?

**Comment by ThomasBrady** [comment]
> @ThomasBrady what other tests are failing?
Thanks for the fix! I saw the residual carryOver but I didn't think of just clearing it 😅 .
I think they are all passing now, I fixed two other test cases today (`generate soroban load`, `overlay flow control`) as well as reading the config from the db. Running a full make check to confirm, then I'll push the update.

**Comment by marta-lokhova** [comment]
@ThomasBrady looks like CI is failing?

**Comment by ThomasBrady** [comment]
> @ThomasBrady looks like CI is failing?

Formatting issues 🙃 re-running now 🙏

**Comment by marta-lokhova** [comment]
@ThomasBrady thanks for the updates! Could we also confirm that acceptance tests pass? You can run `[acceptance]`  tag to verify that.

File: src/overlay/test/OverlayTests.cpp [hunk 5]
```diff
@@ -2141,33 +2141,32 @@ TEST_CASE("overlay flow control", "[overlay][flowcontrol]")
...
-            [&] { return simulation->haveAllExternalized(2, 1); },
-            3 * Herder::EXP_LEDGER_TIMESPAN_SECONDS, false);
         // Generate a bit of load to flood transactions, make sure nodes can
         // close ledgers properly
         auto& loadGen = node->getLoadGenerator();
-        loadGen.generateLoad(
-            GeneratedLoadConfig::createAccountsLoad(/* nAccounts */ 150,
-                                                    /* txRate */ 1));
+        loadGen.generateLoad(GeneratedLoadConfig::createAccountsLoad(
+            /* nAccount */ 150,
```
Comment (marta-lokhova) [inline src/overlay/test/OverlayTests.cpp:2168]: please revert this

File: src/simulation/LoadGenerator.cpp [hunk 6]
```diff
@@ -2183,9 +2220,7 @@ LoadGenerator::execute(TransactionTestFramePtr& txf, LoadGenMode mode,
                 : "";
         CLOG_INFO(LoadGen, "tx rejected '{}': ===> {}, {}",
                   TX_STATUS_STRING[static_cast<int>(addResult.code)],
-                  txf->isSoroban() ? "soroban"
-                                   : xdrToCerealString(txf->getEnvelope(),
-                                                       "TransactionEnvelope"),
+                  xdrToCerealString(txf->getEnvelope(), "TransactionEnvelope"),
```
Comment (marta-lokhova) [inline src/simulation/LoadGenerator.cpp:2223]: I think the problem is with transactions like UPLOAD_WASM, which print the whole multi-KB WASM blob. Ideally, we print the envelope contents without WASM, but that seems outside of the scope of this PR, so I'd just revert this.

File: src/simulation/test/LoadGeneratorTests.cpp [hunk 7]
```diff
@@ -339,13 +376,13 @@ TEST_CASE("generate soroban load", "[loadgen][soroban]")
         }
     }
 
-    // Manually override network settings for invoke load gen
-    for (auto node : nodes)
-    {
-        overrideSorobanNetworkConfigForTest(*node);
-        modifySorobanNetworkConfig(*node, [&](SorobanNetworkConfig& cfg) {
+    // Override network settings for invoke load gen
```
Comment (marta-lokhova) [inline src/simulation/test/LoadGeneratorTests.cpp:379]: Should we update this comment? We're not "overriding" configs anymore, as the network just goes through a normal upgrade flow.

File: src/simulation/test/LoadGeneratorTests.cpp [hunk 8]
```diff
@@ -91,6 +91,49 @@ TEST_CASE("generate load with unique accounts", "[loadgen]")
...
+        });
+
+    simulation->startAllNodes();
+    simulation->crankUntil(
+        [&]() { return simulation->haveAllExternalized(3, 1); },
+        2 * Herder::EXP_LEDGER_TIMESPAN_SECONDS, false);
+    auto nodes = simulation->getNodes();
+    auto& app = *nodes[0]; // pick a node to generate load
+
+    // Manually override network settings for invoke load gen
```
Comment (marta-lokhova) [inline src/simulation/test/LoadGeneratorTests.cpp:110]: Let's update this comment, since we're not manually overriding configs anymore

File: src/simulation/test/LoadGeneratorTests.cpp [hunk 9]
```diff
@@ -339,13 +376,13 @@ TEST_CASE("generate soroban load", "[loadgen][soroban]")
...
-        overrideSorobanNetworkConfigForTest(*node);
-        modifySorobanNetworkConfig(*node, [&](SorobanNetworkConfig& cfg) {
+    // Override network settings for invoke load gen
+    upgradeSorobanNetworkConfig(
+        [&](SorobanNetworkConfig& cfg) {
+            setSorobanNetworkConfigForTest(cfg);
+
             // Entries should never expire
-            cfg.mStateArchivalSettings.maxEntryTTL = 1'000'000;
+            cfg.mStateArchivalSettings.maxEntryTTL = 2'000'000;
```
Comment (marta-lokhova) [inline src/simulation/test/LoadGeneratorTests.cpp:385]: Why this change?

File: src/test/TestUtils.cpp (hunk 1 above)
Comment (marta-lokhova) [reply src/test/TestUtils.cpp:213]: can we track this issue somewhere though? It'd be nice to get rid of account creation, and manipulations with the offset, which are quite error-prone.

File: src/test/TestUtils.cpp [hunk 10]
```diff
@@ -194,6 +197,78 @@ genesis(int minute, int second)
...
+        scheduledUpgrades.mConfigUpgradeSetKey = upgradeSetKey;
+        app->getHerder().setUpgrades(scheduledUpgrades);
+    }
+    // Wait for upgrade to be applied
+    simulation->crankUntil(
+        [&]() {
+            auto netCfg = app.getLedgerManager().getSorobanNetworkConfig();
+            return netCfg == cfg;
+        },
+        300 * Herder::EXP_LEDGER_TIMESPAN_SECONDS, false);
```
Comment (marta-lokhova) [inline src/test/TestUtils.cpp:269]: 300 ledgers seems like a really long timeout, in reality you should see the upgrade going through on the next ledgers, so you can set it to 2 to be safe.

File: src/test/TestUtils.h [hunk 11]
```diff
@@ -16,6 +16,8 @@ namespace stellar
 {
 
 class LoopbackPeerConnection;
+class Simulation;
+using SimulationPtr = std::shared_ptr<Simulation>;
```
Comment (marta-lokhova) [inline src/test/TestUtils.h:20]: Do we need this? SimulationPtr isn't used anywhere outside of this util file

File: src/test/TestUtils.cpp [hunk 12]
```diff
@@ -3,11 +3,14 @@
 // of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
 
 #include "TestUtils.h"
+#include "ledger/LedgerTxn.h"
 #include "overlay/test/LoopbackPeer.h"
+#include "simulation/LoadGenerator.h"
+#include "simulation/Simulation.h"
 #include "test/TxTests.h"
 #include "test/test.h"
+#include "util/StatusManager.h"
```
Comment (marta-lokhova) [inline src/test/TestUtils.cpp:12]: unused include (same with  LedgerTxn)

File: src/herder/test/HerderTests.cpp [hunk 13]
```diff
@@ -50,6 +50,8 @@
 #include <numeric>
 #include <optional>
 
+#include "util/XDRCereal.h"
```
Comment (marta-lokhova) [inline src/herder/test/HerderTests.cpp:53]: unused?

**Review by marta-lokhova — COMMENTED** [review]
thanks, looks good overall! I left a few comments regarding code cleanup.

File: src/simulation/test/LoadGeneratorTests.cpp (hunk 9 above)
Comment (ThomasBrady) [reply src/simulation/test/LoadGeneratorTests.cpp:385]: `SorobanNetworkConfig::isValidConfigSettingEntry` [checks that ](https://github.com/stellar/stellar-core/blob/master/src/ledger/NetworkConfig.cpp#L799) `cfg.stateArchivalSettings().maxEntryTTL >= MinimumSorobanNetworkConfig::MAXIMUM_ENTRY_LIFETIME`, and `MAXIMUM_ENTRY_LIFETIME = 1'054'080;`, so 1 million is an invalid setting. The previously used ltx-based `modifySorobanNetworkConfig` function bypasses the validity check.

File: src/test/TestUtils.cpp (hunk 1 above)
Comment (ThomasBrady) [reply src/test/TestUtils.cpp:213]: https://github.com/stellar/stellar-core/issues/4445

File: src/overlay/test/OverlayTests.cpp [hunk 14]
```diff
@@ -2141,33 +2141,32 @@ TEST_CASE("overlay flow control", "[overlay][flowcontrol]")
         simulation->addPendingConnection(vNode2NodeID, vNode3NodeID);
         simulation->addPendingConnection(vNode3NodeID, vNode1NodeID);
 
-        for (auto& node2 : simulation->getNodes())
+        simulation->startAllNodes();
+        simulation->crankUntil(
```
[the remark this answers is not in the document]
Comment (ThomasBrady) [reply src/overlay/test/OverlayTests.cpp:2145]: I missed this one, checking now

[the remark this answers is not in the document]
Comment (ThomasBrady) [reply src/overlay/test/OverlayTests.cpp]: deleted

File: src/overlay/test/OverlayTests.cpp (hunk 5 above)
Comment (marta-lokhova) [reply src/overlay/test/OverlayTests.cpp]: the typo is still there, should be nAccounts

Comment (ThomasBrady) [reply src/overlay/test/OverlayTests.cpp]: oops I thought you meant the formatting 😅 done.

**Review by marta-lokhova — APPROVED** [review]
